src/ipred_prepare.rs

```diff
@@ -301,53 +285,48 @@ pub unsafe fn rav1d_prepare_intra_edges<BD: BitDepth>(
                 (edge_flags & EDGE_I444_TOP_HAS_RIGHT) != 0
             };
             if have_topright {
+                let top_right = &mut top[sz_0 as usize..];
                 let px_have_2 = cmp::min(sz_0, w - x - tw << 2);
-                memcpy(
-                    top.offset(sz_0 as isize) as *mut c_void,
-                    dst_top.offset(sz_0 as isize) as *const c_void,
-                    (px_have_2 << 1) as usize,
-                );
+                for i in 0..px_have_2 {
+                    top_right[i as usize] = *dst_top.offset((sz_0 + i) as isize);
+                }
                 if px_have_2 < sz_0 {
+                    let fill_value = top_right[px_have_2 as usize - 1];
                     BD::pixel_set(
-                        slice::from_raw_parts_mut(
-                            top.offset(sz_0 as isize).offset(px_have_2 as isize),
-                            (sz_0 - px_have_2).try_into().unwrap(),
-                        ),
-                        *top.offset((sz_0 + px_have_2 - 1) as isize),
-                        (sz_0 - px_have_2).try_into().unwrap(),
+                        &mut top_right[px_have_2 as usize..],
+                        fill_value,
+                        (sz_0 - px_have_2) as usize,
                     );
                 }
             } else {
-                BD::pixel_set(
-                    slice::from_raw_parts_mut(top.offset(sz_0 as isize), sz_0.try_into().unwrap()),
-                    *top.offset((sz_0 - 1) as isize),
-                    sz_0.try_into().unwrap(),
-                );
+                let fill_value = top[sz_0 as usize - 1];
+                let top_right = &mut top[sz_0 as usize..];
+                BD::pixel_set(top_right, fill_value, sz_0 as usize);
             }
         }
     }
     if av1_intra_prediction_edges[mode as usize]
         .needs
         .contains(Needs::TOP_LEFT)
     {
+        // top-left sample and immediate neighbours
+        let corner = &mut topleft_out[(topleft_origin - 1)..(topleft_origin + 2)];
```

Comment:
Your way works and is good, but if we make `corner` an array we eliminate bounds checking, like this:

```suggestion
        let corner = <&mut [_; 3]>::try_from(&mut topleft_out[topleft_origin - 1..][..3]).unwrap();
```

Reply:
If one defines a sub-slice of size 3 such as `&mut data[first_idx..][..3]`, can't the compiler figure out that bounds checking should be done only once (i.e. the sub-slice fits within the slice) and not every time the sub-slice is accessed if the only indices used are 0, 1 and 2?

Reply:
It can but it can be tripped up sometimes.  With an array that optimization is guaranteed.

Reply:
Thanks for the clarification. I made the change as suggested